### PR TITLE
7.x.owner proxies

### PR DESCRIPTION
**Pull request by willtp87**

### CODE REVIEW

File: islandora_owner_permissions.module [hunk 1]
```diff
@@ -68,13 +123,23 @@ function islandora_owner_permissions_module_implements_alter(&$implementations,
...
-  if (!isset($user->name) || $user->name != $object->owner || !array_key_exists($op, $perm_map)) {
+  if (($user->name != $object->owner && !array_key_exists($user->uid, $proxies)) ||
+    !(array_key_exists($op, $perm_map) && array_key_exists($op, $proxy_perm_map))) {
     return islandora_islandora_object_access($op, $object, $user);
   }
   else {
-    return user_access($perm_map[$op], $user);
+    return (
+      ($user->name == $object->owner && user_access($perm_map[$op], $user)) ||
+      (array_key_exists($user->uid, $proxies) && user_access($proxy_perm_map[$op], $user))
```
Comment (adam-vessey) [inline islandora_owner_permissions.module]: `s/user_access/islandora_object_access/`?

File: islandora_owner_permissions.module [hunk 2]
```diff
@@ -83,12 +148,42 @@ function islandora_owner_permissions_islandora_object_access($op, $object, $user
  */
 function islandora_owner_permissions_islandora_datastream_access($op, $datastream, $user) {
   module_load_include('inc', 'islandora_owner_permissions', 'includes/utilities');
+  module_load_include('inc', 'islandora_owner_permissions', 'includes/db');
+  if (!isset($user->name) || !isset($user->uid)) {
+    return user_access($perm_map[$op], $user);
```
Comment (adam-vessey) [inline islandora_owner_permissions.module]: `s/user_access/islandora_object_access/`?

File: islandora_owner_permissions.module [hunk 3]
```diff
@@ -83,12 +148,42 @@ function islandora_owner_permissions_islandora_object_access($op, $object, $user
...
-  if (!isset($user->name) || $user->name != $datastream->parent->owner || !array_key_exists($op, $perm_map)) {
+  if (($user->name != $datastream->parent->owner && !array_key_exists($user->uid, $proxies)) ||
+    !(array_key_exists($op, $perm_map) && array_key_exists($op, $proxy_perm_map))) {
     return islandora_islandora_datastream_access($op, $datastream, $user);
   }
   else {
-    return user_access($perm_map[$op], $user);
+    return (
+      ($user->name == $datastream->parent->owner && user_access($perm_map[$op], $user)) ||
+      (array_key_exists($user->uid, $proxies) && user_access($proxy_perm_map[$op], $user))
```
Comment (adam-vessey) [inline islandora_owner_permissions.module]: `s/user_access/islandora_object_access/`?

File: islandora_owner_permissions.module [hunk 4]
```diff
@@ -68,13 +123,23 @@ function islandora_owner_permissions_module_implements_alter(&$implementations,
  */
 function islandora_owner_permissions_islandora_object_access($op, $object, $user) {
   module_load_include('inc', 'islandora_owner_permissions', 'includes/utilities');
+  module_load_include('inc', 'islandora_owner_permissions', 'includes/db');
+  if (!isset($user->name) || !isset($user->uid)) {
+    return user_access($perm_map[$op], $user);
```
Comment (adam-vessey) [inline islandora_owner_permissions.module]: `s/user_access/islandora_object_access/`?

File: islandora_owner_permissions.module [hunk 5]
```diff
@@ -68,13 +123,23 @@ function islandora_owner_permissions_module_implements_alter(&$implementations,
...
   // Call original if we are not the owner or don't override the op.
-  if (!isset($user->name) || $user->name != $object->owner || !array_key_exists($op, $perm_map)) {
+  if (($user->name != $object->owner && !array_key_exists($user->uid, $proxies)) ||
+    !(array_key_exists($op, $perm_map) && array_key_exists($op, $proxy_perm_map))) {
     return islandora_islandora_object_access($op, $object, $user);
   }
   else {
-    return user_access($perm_map[$op], $user);
+    return (
+      ($user->name == $object->owner && user_access($perm_map[$op], $user)) ||
```
Comment (adam-vessey) [inline islandora_owner_permissions.module]: `s/user_access/islandora_object_access/`?

File: islandora_owner_permissions.module [hunk 6]
```diff
@@ -83,12 +148,42 @@ function islandora_owner_permissions_islandora_object_access($op, $object, $user
...
   // Call original if we are not the owner or don't override the op.
-  if (!isset($user->name) || $user->name != $datastream->parent->owner || !array_key_exists($op, $perm_map)) {
+  if (($user->name != $datastream->parent->owner && !array_key_exists($user->uid, $proxies)) ||
+    !(array_key_exists($op, $perm_map) && array_key_exists($op, $proxy_perm_map))) {
     return islandora_islandora_datastream_access($op, $datastream, $user);
   }
   else {
-    return user_access($perm_map[$op], $user);
+    return (
+      ($user->name == $datastream->parent->owner && user_access($perm_map[$op], $user)) ||
```
Comment (adam-vessey) [inline islandora_owner_permissions.module]: `s/user_access/islandora_object_access/`?

File: includes/db.inc [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  return $id;
+}
+
+/**
+ * Creates/gets the database id related to a PID.
+ *
+ * @throws Exception if the object cannot be loaded.
+ */
+function islandora_owner_permissions_pid_to_db($pid) {
+  $object = islandora_object_load($pid);
```
Comment (adam-vessey) [inline includes/db.inc]: Not too terribly concerned, but do we particularly care about this?... Could hypothetically be useful to call before the object is loadable, say, from an ingest step.

File: includes/db.inc [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ * Remove an object from the db.
+ */
+function islandora_owner_permissions_remove_object_from_db($pid) {
+  $pid_id = islandora_owner_permissions_get_pid_id($pid);
+  db_delete('islandora_owner_permissions_objects')
+    ->condition('id', $pid_id)
+    ->execute();
+  db_delete('islandora_owner_permissions_object_proxies')
+    ->condition('object', $pid_id)
+    ->execute();
```
Comment (adam-vessey) [inline includes/db.inc:65]: Want to do the `db_transaction()` thing here?

**Review by adam-vessey — COMMENTED** [review]
Couple of things...

File: includes/db.inc (hunk 7 above)
Comment (willtp87) [reply includes/db.inc]: I would like to avoid it.

File: includes/db.inc (hunk 8 above)
Comment (adam-vessey) [reply includes/db.inc:65]: Is did.

File: includes/db.inc (hunk 7 above)
Comment (adam-vessey) [reply includes/db.inc]: Okay.

**Pull request merged**